src/Workspaces/Core/Portable/SemanticModelReuse/SemanticModelWorkspaceServiceFactory.SemanticModelWorkspaceService.cs

```diff
@@ -0,0 +1,175 @@
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.SemanticModelReuse
+{
+    internal readonly struct SemanticModelReuseInfo
+    {
+        /// <summary>
+        /// The original *non-speculative* semantic model we retrieved for this document at some point.
+        /// </summary>
+        public readonly SemanticModel PreviousNonSpeculativeSemanticModel;
+
+        /// <summary>
+        /// The current semantic model we retrieved <see cref="SemanticModel"/> for the <see cref="BodyNode"/>.  Could
+        /// be speculative or non-speculative.
+        /// </summary>
+        public readonly SemanticModel CurrentSemanticModel;
+
+        /// <summary>
+        /// The current method body we retrieved the <see cref="CurrentSemanticModel"/> for.
+        /// </summary>
+        public readonly SyntaxNode BodyNode;
+
+        /// <summary>
+        /// The top level version of the project when we retrieved <see cref="SemanticModel"/>.  As long as this is the
+        /// same we can continue getting speculative models to use.
+        /// </summary>
+        public readonly VersionStamp TopLevelSementicVersion;
+
+        public SemanticModelReuseInfo(SemanticModel previousNonSpeculativeSemanticModel, SemanticModel currentSemanticModel, SyntaxNode bodyNode, VersionStamp topLevelSementicVersion)
+        {
+            PreviousNonSpeculativeSemanticModel = previousNonSpeculativeSemanticModel;
+            CurrentSemanticModel = currentSemanticModel;
+            BodyNode = bodyNode;
+            TopLevelSementicVersion = topLevelSementicVersion;
+        }
+    }
+
+    internal partial class SemanticModelReuseWorkspaceServiceFactory : IWorkspaceServiceFactory
+    {
+        private sealed class SemanticModelReuseWorkspaceService : ISemanticModelReuseWorkspaceService
+        {
+            /// <summary>
+            /// A mapping from a document id to the last semantic model we produced for it, along with the top level
+            /// semantic version that that semantic model corresponds to.  We can continue reusing the semantic model as
+            /// long as no top level changes occur.
+            /// <para/>
+            /// In general this dictionary will only contain a single key-value pair.  However, in the case of linked
+            /// documents, there will be a key-value pair for each of the independent document links that a document
+            /// has.
+            /// </summary>
+            private ImmutableDictionary<DocumentId, SemanticModelReuseInfo?> _semanticModelMap = ImmutableDictionary<DocumentId, SemanticModelReuseInfo?>.Empty;
```

Comment:
A bit worried here that nothing empties this on solution close -- if you just close a solution in VS we're still rooting all the Compilations/SemanticModels; it'll get cleared if you open and edit something else as long as it's C#/VB, but otherwise it's a leak for good if you switch to some C++ code.

Reply:
yes.  that's a fair point.  i will listen for the event here.

Reply:
done.